### PR TITLE
[LVI] Handle icmp of ashr.

### PR DESCRIPTION
**Pull request by aemerson**
This handles the case where this combine:
icmp sgt (ashr X, ShAmtC), C --> icmp sgt X, ((C + 1) << ShAmtC) - 1

wasn't performed by instcombine.

Proof of the original combine: https://alive2.llvm.org/ce/z/SfpsvX

This is a port of the review in https://reviews.llvm.org/D151911 to GitHub.

### CODE REVIEW

**Comment by aemerson** [comment]
ping @nikic

**Comment by aemerson** [comment]
ping @nikic I've been trying to get this in for a few months now. Should I just give up on this change? If so, I'd like to go ahead with https://reviews.llvm.org/D149918 so that we can re-commit the always-inliner change we had to back out early this year.

**Review by nikic — COMMENTED** [review]
This doesn't compute the correct ranges for slt and sge predicates, because the +/-1 adjustment needs to happen on C prior to shifting, not the result.

I gave my suggestion from the Phabricator review a try with a patch like this: https://gist.github.com/nikic/e040c8870eb6d4aa59057b332ff1ce40 I had to adjust some of the test expectations, but I believe that's because they weren't correct due to the issue above.

File: llvm/lib/Analysis/LazyValueInfo.cpp [hunk 1]
```diff
@@ -1148,6 +1149,57 @@ static ValueLatticeElement getValueFromICmpCondition(Value *Val, ICmpInst *ICI,
...
+      llvm_unreachable("Unknown signed predicate!");
+    case ICmpInst::ICMP_SGT:
+      Lower = New + 1;
+      Upper = MaxSigned;
+      break;
+    case ICmpInst::ICMP_SLE:
+      Lower = MinSigned;
+      Upper = New + 1;
+      break;
+    case ICmpInst::ICMP_SGE:
```
Comment (goldsteinn) [inline llvm/lib/Analysis/LazyValueInfo.cpp:1189]: Does this hold for preds other then SGT/SLE: https://alive2.llvm.org/ce/z/wHF_sk

**Comment by aemerson** [comment]
> This doesn't compute the correct ranges for slt and sge predicates, because the +/-1 adjustment needs to happen on C prior to shifting, not the result.
> 
> I gave my suggestion from the Phabricator review a try with a patch like this: https://gist.github.com/nikic/e040c8870eb6d4aa59057b332ff1ce40 I had to adjust some of the test expectations, but I believe that's because they weren't correct due to the issue above.

Thanks, I've merged your patch into this.

**Review by nikic — COMMENTED** [review]
Basically okay, but I think the constants in the tests need to be adjusted from `icmp sge i8 %s, 0` to some other value like `icmp sge i8 %s, 1`. The zero constant is a bad choice because shifting zero gives you zero again, so the correct adjustment of the constant isn't tested for all predicates.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**